src/handle.c

```diff
@@ -136,12 +142,18 @@ SEXP R_handle_setopt(SEXP ptr, SEXP keys, SEXP values){
     if(val == R_NilValue){
       assert(curl_easy_setopt(handle, key, NULL));
     } else if (key == CURLOPT_PROGRESSFUNCTION) {
-      if (TYPEOF(val) != CLOSXP)
-        error("Value for option %s (%d) must be a function.", optname, key);
-
-      assert(curl_easy_setopt(handle, CURLOPT_PROGRESSFUNCTION,
-        (curl_progress_callback) R_curl_callback_progress));
-      assert(curl_easy_setopt(handle, CURLOPT_PROGRESSDATA, val));
+      if (TYPEOF(val) != CLOSXP || ! isNull(val))
+        error("Value for option %s (%d) must be a function or NULL.", optname, key);
+
+      /* Disable or replace a curl progress callback */
+      if(isNull(val))
```

Comment:
I agree, readability is poor here. I will refactor this sequence of expressions.